File: server/src/datasource/file_system_item.py
from enum import Enum
from typing import List
from datetime import datetime


class FileSystemItemType(Enum):
    DIRECTORY = "directory"
    FILE = "file"


class FileSystemItem:
    def __init__(
        self,
        id: str,
        name: str,
        type: FileSystemItemType,
        parent_id: str,
        path: str,
        created_at: datetime,
        updated_at: datetime,
        tags: List[str],
    ):
        self.id = id
        self.name = name
        self.type = type
        self.parent_id = parent_id
        self.path = path
        self.created_at = created_at
        self.updated_at = updated_at
        self.tags = tags

    @property
    def is_directory(self) -> bool:
        return self.type == FileSystemItemType.DIRECTORY

    @property
    def is_file(self) -> bool:
        return self.type == FileSystemItemType.FILE

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "type": self.type.value,
            "parent_id": self.parent_id,
            "path": self.path,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
            "tags": self.tags,
        }

    @classmethod
    def from_dict(cls, data: dict):
        return cls(
            id=data["id"],
            name=data["name"],
            type=FileSystemItemType(data["type"]),
            parent_id=data["parent_id"],
            path=data["path"],
            created_at=datetime.fromisoformat(data["created_at"]),
            updated_at=datetime.fromisoformat(data["updated_at"]),
            tags=data["tags"],
        )

File: flutter/lib/src/file_system/file_system_item.dart
import 'package:flutter/material.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:uuid/uuid.dart';

part 'file_system_item.g.dart';

enum FileSystemItemType {
  directory,
  pdf,
  link,
}

@JsonSerializable(explicitToJson: true, fieldRename: FieldRename.snake)
class FileSystemItem {
  final String id;
  final String name;
  final FileSystemItemType type;
  final String parentId;
  final String path;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<String> tags;
  final bool processed;
  final String? url;
  final String? fsId; // the id of the document in the file system
  final String? indexId;

  FileSystemItem({
    required this.id,
    required this.name,
    required this.type,
    required this.parentId,
    required this.path,
    required this.createdAt,
    required this.updatedAt,
    required this.tags,
    this.processed = false,
    this.url,
    this.fsId,
    this.indexId,
  });

  bool get isDirectory => type == FileSystemItemType.directory;
  bool get isPdf => type == FileSystemItemType.pdf;
  bool get isLink => type == FileSystemItemType.link;
  bool get isRoot => (type == FileSystemItemType.directory && id == '-1');

  static FileSystemItem createFromAnotherFileSystemItem(
    FileSystemItem anotherItem, {
    required String name,
    required FileSystemItemType type,
    required List<String> tags,
    String? url,
    String? fsId,
    String? indexId,
  }) {
    String id = const Uuid().v4();
    String parentId =
        !anotherItem.isDirectory ? anotherItem.parentId : anotherItem.id;
    String path = !anotherItem.isDirectory
        ? "${anotherItem.path.substring(0, anotherItem.path.lastIndexOf("/"))}/$id"
        : "${anotherItem.path}/$id";

    return FileSystemItem(
      id: id,
      name: name,
      type: type,
      parentId: parentId,
      path: path,
      createdAt: DateTime.now(),
      updatedAt: DateTime.now(),
      tags: tags,
      processed: false,
      url: url,
      fsId: fsId,
      indexId: indexId,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is FileSystemItem &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;

  factory FileSystemItem.fromJson(Map<String, dynamic> json) =>
      _$FileSystemItemFromJson(json);

  Map<String, dynamic> toJson() => _$FileSystemItemToJson(this);

  static FileSystemItem getRootItem() {
    return FileSystemItem(
      id: '-1',
      name: 'Root',
      type: FileSystemItemType.directory,
      parentId: '',
      path: '',
      createdAt: DateTime.now(),
      updatedAt: DateTime.now(),
      tags: ['root'],
      processed: true,
    );
  }

  static Icon getIconForFileSystemItem(FileSystemItem item,
      {bool isOutlined = false}) {
    IconData iconData;
    switch (item.type) {
      case FileSystemItemType.directory:
        iconData = isOutlined ? Icons.folder_outlined : Icons.folder;
        break;
      case FileSystemItemType.pdf:
        iconData =
            isOutlined ? Icons.picture_as_pdf_outlined : Icons.picture_as_pdf;
        break;
      case FileSystemItemType.link:
        iconData = isOutlined ? Icons.link : Icons.link;
        break;
      default:
        iconData = isOutlined
            ? Icons.insert_drive_file_outlined
            : Icons.insert_drive_file;
        break;
    }
    return Icon(iconData);
  }
}
